### PR TITLE
refactor: compile args + config to opts

### PR DESCRIPTION
**Pull request by NicholasLYang**
### Description

Refactors our config setup to explicitly transform the `Args` and `Config` to `Opts` in the `CommandBase` constructor. This accomplishes a few things:

- Limits our usage of args beyond `cli`. Args are structured according to the CLI parsing logic, so there are multiple fields that do the same thing (args.command.run_args and args.run_args). By compiling down to `Opts`, we can use a more normalized structure.
- Remove the requirement that we have a `Command::Run` to create a `Run` struct. This requires us to mock out a `Command::Run` in places like `query` or `ls`, which is unnecessary.
- Fix a potential nasty bug where we change the `Args` struct via `CommandBase::args_mut` but that change is not propagated to the config because it's cached via a `OnceCell`.
- Finally, and most importantly, allow us to construct a `Run` without `Args`. This is very important for `query`.

However, this does create a change where `CommandBase::new` is now fallible. Overall this is fine, but it does mean that we sometimes error earlier than before. I couldn't figure out an alternative without making `Config` reactive on any `Args` change

### CODE REVIEW

**Review by chris-olszewski — CHANGES_REQUESTED** [review]
Great step to cleaning up this logic.

Only blocker is the change to `turbo ls` flag behavior.

File: crates/turborepo-lib/src/cli/mod.rs [hunk 1]
```diff
@@ -1258,49 +1261,44 @@ pub async fn run(
...
+            let base = CommandBase::new(cli_args.clone(), repo_root, version, color_config)?;
             config::run(base).await?;
             Ok(0)
         }
         Command::Ls {
-            affected,
-            filter,
-            packages,
-            output,
+            packages, output, ..
```
[the remark this answers is not in the document]
Comment (NicholasLYang) [reply crates/turborepo-lib/src/cli/mod.rs:1291]: If I did this correctly, I think I moved the logic [here](https://github.com/vercel/turborepo/pull/9552/files#diff-c6390e97c21ab30e6f307b7fbaa5a25154179d1fa544b25ea1dcd2233c3a7902R108).

[the remark this answers is not in the document]
Comment (chris-olszewski) [reply crates/turborepo-lib/src/cli/mod.rs:1291]: Huh, looks like this is an existing bug? Not sure what's happening here:
```
[1 olszewski@chriss-mbp] /Users/olszewski/code/vercel/turborepo $ turbo --version    
 WARNING  No locally installed `turbo` found. Using version: 2.3.4-canary.0.
2.3.4-canary.0
[0 olszewski@chriss-mbp] /Users/olszewski/code/vercel/turborepo $ turbo ls --affected
 WARNING  No locally installed `turbo` found. Using version: 2.3.4-canary.0.
turbo 2.3.4-canary.0

 WARNING  ls command is experimental and may change in the future
0 no packages (pnpm)

[0 olszewski@chriss-mbp] /Users/olszewski/code/vercel/turborepo $ turbo --affected ls           
 WARNING  No locally installed `turbo` found. Using version: 2.3.4-canary.0.
turbo 2.3.4-canary.0

 WARNING  ls command is experimental and may change in the future
37 packages (pnpm)

  @turbo-internal/top-issues-gh-action packages/top-issues
  @turbo/benchmark packages/turbo-benchmark
  @turbo/codemod packages/turbo-codemod
  @turbo/eslint-config packages/eslint-config
  @turbo/exe-stub packages/turbo-exe-stub
  @turbo/gen packages/turbo-gen
  @turbo/telemetry packages/turbo-telemetry
  @turbo/test-utils packages/turbo-test-utils
  @turbo/tsconfig packages/tsconfig
  @turbo/types packages/turbo-types
  @turbo/utils packages/turbo-utils
  @turbo/workspaces packages/turbo-workspaces
  @turborepo-examples-tests/basic turborepo-tests/example-basic
  @turborepo-examples-tests/design-system turborepo-tests/example-design-system
  @turborepo-examples-tests/kitchen-sink turborepo-tests/example-kitchen-sink
  @turborepo-examples-tests/non-monorepo turborepo-tests/example-non-monorepo
  @turborepo-examples-tests/with-berry turborepo-tests/example-with-berry
  @turborepo-examples-tests/with-changesets turborepo-tests/example-with-changesets
  @turborepo-examples-tests/with-npm turborepo-tests/example-with-npm
  @turborepo-examples-tests/with-react-native-web turborepo-tests/example-with-react-native-web
  @turborepo-examples-tests/with-rollup turborepo-tests/example-with-rollup
  @turborepo-examples-tests/with-svelte turborepo-tests/example-with-svelte
  @turborepo-examples-tests/with-tailwind turborepo-tests/example-with-tailwind
  @turborepo-examples-tests/with-vite turborepo-tests/example-with-vite
  @turborepo-examples-tests/with-yarn turborepo-tests/example-with-yarn
  cargo-sweep-action .github/actions/cargo-sweep
  cli cli
  create-turbo packages/create-turbo
  eslint-config-turbo packages/eslint-config-turbo
  eslint-plugin-turbo packages/eslint-plugin-turbo
  prysk packages/prysk
  turbo-ignore packages/turbo-ignore
  turbo-vsc packages/turbo-vsc
  turborepo-examples examples
  turborepo-repository packages/turbo-repository
  turborepo-tests-helpers turborepo-tests/helpers
  turborepo-tests-integration turborepo-tests/integration
[0 olszewski@chriss-mbp] /Users/olszewski/code/vercel/turborepo $ turbo_dev ls --affected
 WARNING  No locally installed `turbo` found. Using version: 2.3.3.
turbo 2.3.3

 WARNING  ls command is experimental and may change in the future
0 no packages (pnpm)

[0 olszewski@chriss-mbp] /Users/olszewski/code/vercel/turborepo $ turbo_dev --affected ls           
 WARNING  No locally installed `turbo` found. Using version: 2.3.3.
turbo 2.3.3

 WARNING  ls command is experimental and may change in the future
37 packages (pnpm)

  @turbo-internal/top-issues-gh-action packages/top-issues
  @turbo/benchmark packages/turbo-benchmark
  @turbo/codemod packages/turbo-codemod
  @turbo/eslint-config packages/eslint-config
  @turbo/exe-stub packages/turbo-exe-stub
  @turbo/gen packages/turbo-gen
  @turbo/telemetry packages/turbo-telemetry
  @turbo/test-utils packages/turbo-test-utils
  @turbo/tsconfig packages/tsconfig
  @turbo/types packages/turbo-types
  @turbo/utils packages/turbo-utils
  @turbo/workspaces packages/turbo-workspaces
  @turborepo-examples-tests/basic turborepo-tests/example-basic
  @turborepo-examples-tests/design-system turborepo-tests/example-design-system
  @turborepo-examples-tests/kitchen-sink turborepo-tests/example-kitchen-sink
  @turborepo-examples-tests/non-monorepo turborepo-tests/example-non-monorepo
  @turborepo-examples-tests/with-berry turborepo-tests/example-with-berry
  @turborepo-examples-tests/with-changesets turborepo-tests/example-with-changesets
  @turborepo-examples-tests/with-npm turborepo-tests/example-with-npm
  @turborepo-examples-tests/with-react-native-web turborepo-tests/example-with-react-native-web
  @turborepo-examples-tests/with-rollup turborepo-tests/example-with-rollup
  @turborepo-examples-tests/with-svelte turborepo-tests/example-with-svelte
  @turborepo-examples-tests/with-tailwind turborepo-tests/example-with-tailwind
  @turborepo-examples-tests/with-vite turborepo-tests/example-with-vite
  @turborepo-examples-tests/with-yarn turborepo-tests/example-with-yarn
  cargo-sweep-action .github/actions/cargo-sweep
  cli cli
  create-turbo packages/create-turbo
  eslint-config-turbo packages/eslint-config-turbo
  eslint-plugin-turbo packages/eslint-plugin-turbo
  prysk packages/prysk
  turbo-ignore packages/turbo-ignore
  turbo-vsc packages/turbo-vsc
  turborepo-examples examples
  turborepo-repository packages/turbo-repository
  turborepo-tests-helpers turborepo-tests/helpers
  turborepo-tests-integration turborepo-tests/integration
```

**Review by chris-olszewski — APPROVED** [review]
Since it's an already existing bug, we can 🚢 , but very confused as to what's happening. Repro by checking out this branch and using `turbo ls --affected`

**Pull request merged**